util/deadlockhistory/deadlock_history.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deadlockhistory
+
+import (
+	"encoding/hex"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pingcap/parser/mysql"
+	tikverr "github.com/pingcap/tidb/store/tikv/error"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/resourcegrouptag"
+	"go.uber.org/zap"
+)
+
+// WaitChainItem represents an entry in a deadlock's wait chain.
+type WaitChainItem struct {
+	TryLockTxn     uint64
+	SQLDigest      string
+	Key            []byte
+	SQLs           []string
+	TxnHoldingLock uint64
+}
+
+// DeadlockRecord represents a deadlock events, and contains multiple transactions' information.
+type DeadlockRecord struct {
+	// The ID don't need to be set manually and it's set when it's added into the DeadlockHistory by invoking its Push
+	// method.
+	ID        uint64
+	OccurTime time.Time
+	WaitChain []WaitChainItem
+}
+
+// DeadlockHistory is a collection for maintaining recent several deadlock events.
+type DeadlockHistory struct {
+	sync.RWMutex
+
+	deadlocks []*DeadlockRecord
+
+	// The `head` and `size` makes the `deadlocks` array behaves like a deque. The valid elements are
+	// deadlocks[head:head+size], or deadlocks[head:] + deadlocks[:head+size-len] if `head+size` exceeds the array's
+	// length.
+	head int
+	size int
+
+	// currentID is used to allocate IDs for deadlock records pushed to the queue that's unique in the deadlock
+	// history queue instance.
+	currentID uint64
+}
+
+// NewDeadlockHistory creates an instance of DeadlockHistory
+func NewDeadlockHistory(capacity int) *DeadlockHistory {
+	return &DeadlockHistory{
+		deadlocks: make([]*DeadlockRecord, capacity),
+		currentID: 1,
+	}
+}
+
+// GlobalDeadlockHistory is the global instance of DeadlockHistory, which is used to maintain recent several recent
+// deadlock events globally.
+// TODO: Make the capacity configurable
+var GlobalDeadlockHistory = NewDeadlockHistory(10)
```

Comment:
Do we have a correspond deveopment task for the memory management of the history memory buffer?

Reply:
Hmmm, currently not. I think it's better to be a configurable size, but I think we do not need any more memory managements since the capacity is fixed without manual configuration changing. Strictly speaking, its possible that there are some deadlock incidents with really-long wait chains or extremely-long keys, which I think is very unlikely to happen.